### ISSUE TITLE
Materials support at job/stage level

### ISSUE THREAD

**Issue by adityabyreddy** [gocd/gocd#3076]
##### Issue Type

- Feature proposal
- Feature enhancement

##### Summary

Currently, only pipeline have material support. Suppose if a pipeline uses multiple materials, then for every job these materials should be cloned at agents end which affect space and time constraints. If we provide material support at STAGE/JOB level then only required materials will be cloned at agent end and improves the job time. 

##### Environment

There are 3 jobs (J1, J2, J3) configured within a pipeline, doing different things. 3 materials were configured namely project-1, project-2, project-3. While executing J1, both project-2 and project-3 are also cloned which is unnecessary. 

###### Basic environment details

* Go Version: 16.12.0

##### Steps to Reproduce

1. Create a pipeline with 3 different materials(repositories)
2. Create 3 jobs under default stage.
3. Create tasks in individual jobs w.r.t individual repositories.
4. Execute the pipeline.
5. Take a look at any of the job console. 

##### Expected Results

Only the necessary material should be cloned.

##### Actual Results

All the materials configured with pipeline are cloned.

**Comment by arvindsv** [comment]
@adityabyreddy: I understand what you're saying, but I feel that this would further complicate an already complex model. I think people find it hard to understand the pipeline, stage, job, task model itself, and how they related to materials. Including support for materials at multiple levels gets even more confusing. So, I'm going to close this as "won't fix". Thanks!

**Comment by FredrikWendt** [comment]
An option could be to run material fetching tasks in parallel.

**Comment by Prabhakar-cg** [comment]
i am also look something similar to this ask. 
i am having pipeline which consists of 4 stages where as all the stages uses different materials for eg..some are git and some other pipelines.
i dont want to declare all materials in pipeline because my second and 3rd stage is depend on some other pipeline output.
